### PR TITLE
HZN-372: jnxOperatingTable datacollection not correctly made

### PR DESCRIPTION
**Pull request by StCyr**
JIRA: http://issues.opennms.org/browse/HZN-372
Bamboo: http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS90

Created a new resourceType named "jnxOperatingTable" for juniper's snmp data collection.
Also, changed the juniper-router group to make use of this resourceType.

Caution, this PR is not working atm. But, I need your help to go further.

1st: 

I wonder if changing the juniper-router group is the correct way to go. Indeed, it will probably break current data collection of this group. So, maybe I should create a brand new group for this collection, and leave both groups active for a moment. Also the jnxOperatingTable is also available for juniper switches, so the name "juniper-router" is a little bit misleading.

2nd: 

The collection simply doesn't work atm. I've traced down the issue to be due to the match-expression not matching. Ie: I keep getting "o.o.n.c.PersistRegexSelectorStrategy: shouldPersist: checking #jnxOpDescr matches 'Routing Engine' ? false" logs. But I have no idea why.

### CODE REVIEW

**Comment by StCyr** [comment]
As I've said on http://issues.opennms.org/browse/HZN-372, adding the following line in PersistRegexSelectorStrategy.visitAttribute() would help in debugging this persistence selector:

{code}
public void visitAttribute(CollectionAttribute attribute) {
if (StringAttributeType.supportsType(attribute.getType()))
context.setVariable(attribute.getName(), attribute.getStringValue());
+++ LOG.debug("visitAttribute: storing value {} for attribute named {}", attribute.getStringValue(), attribute.getName());
}
{code}

This is something that's done in ThresholdingVisitor.visitAttribute().

Could some developer do that?

**Comment by StCyr** [comment]
Ok, I could get the datacollection working.

It seems like the match-expression's value needs to be surrounded by parenthesis.

Anyway, adding the log line in PersistRegexSelectorStrategy.visitAttribute() could be a good thing IMHO.

**Comment by StCyr** [comment]
I'm now writing the graph defintions but it seems there's a bug in JRobinRrdStrategy.java

It seems it doesn't support blanks in DEF.

In the logs of the graph creation I've the following line:

{code}
2015-07-17 12:09:49,986 DEBUG [qtp823241375-2562 - /opennms/graph/graph.png?resourceId=node%5b12%5d.jnxOperatingTable%5bnode0%2bRouting%2bEngine%5d&report=juniper.jnxOpTableCPU&start=1437041389421&end=1437127789421&width=1085&height=326] o.o.n.r.j.JRobinRrdStrategy: command array = [--imgformat, PNG, --font, DEFAULT:7, --font, TITLE:10, --start, 1437041389, --end, 1437127789, --title="Juniper: CPU usage on node0 Routing Engine", --vertical-label="percent", --lower-limit, 0, --upper-limit, 105, DEF:dpercent=snmp/12/jnxOperatingTable/node0, Routing, Engine/jnxOpCpu.jrb:jnxOpCpu:AVERAGE, LINE2:dpercent#0000ff:"Slot FEB Buffer ", GPRINT:dpercent:AVERAGE:" Avg  \: %8.2lf %s\n", --width, 1085, --height, 326]
{code}

It's putting commas in the DEF attribute:

{code}
DEF:dpercent=snmp/12/jnxOperatingTable/node0, Routing, Engine/jnxOpCpu.jrb:jnxOpCpu:AVERAGE
{code}

**Comment by StCyr** [comment]
I fixed the problem with the JRobinRrdStrategy by adding a replace-all parameter in the SiblingColumnStorageStrategy; It removes all blank characters from the sibling-column-name

**Comment by StCyr** [comment]
Added graph definition for the collected data

**Comment by StCyr** [comment]
Patch looks now good to me.

You can merge it if that's ok for you to modify the "juniper-router" group directly.

If not just tell me and I'll adapt the PR.

Cyr

**Comment by indigo423** [comment]
I let it run through bamboo and the build fails see http://bamboo.internal.opennms.com:8085/browse/OPENNMS-ONMS90-2#new-failed-tests

**Comment by StCyr** [comment]
Hi @indigo423 

I'm off 'till Wednesday. I'll have a look at that time.

Cyr

**Comment by indigo423** [comment]
@StCyr 
I've found one issue regarding the failing tests, fixed it and send you a pull request, so can see what the problem was. Bamboo is crunching again and see what happens next :) 

Thanks a lot for working on this data collection and making Juniper users happier :+1:

**Comment by StCyr** [comment]
Your fix should make bamboo happy 

Thanks 

Cyr

File: opennms-base-assembly/src/main/filtered/etc/snmp-graph.properties.d/juniper-graph.properties [hunk 1]
```diff
@@ -24,6 +28,108 @@ ive.connections
...
+ GPRINT:float5:MAX:"Max \\: %10.2lf\\n" \
+ LINE2:float15#a40000:"15 minute" \
+ GPRINT:float15:AVERAGE:"Avg \\: %10.2lf" \
+ GPRINT:float15:MIN:"Min \\: %10.2lf" \
+ GPRINT:float15:MAX:"Max \\: %10.2lf\\n"
+
+report.juniper.jnxOpTableBuff.name=Juniper Buffer Utilization
+report.juniper.jnxOpTableBuff.columns=jnxOpBuff
+report.juniper.jnxOpTableBuff.type=jnxOperatingTable
+report.juniper.jnxOpTableLoad.propertiesValues=jnxOpDescr
```
Comment (StCyr) [inline opennms-base-assembly/src/main/filtered/etc/snmp-graph.properties.d/juniper-graph.properties]: Hi @indigo423 

I overlooked the following typo in jnxOpTableBuff graph:

+report.juniper.jnxOpTableBuff.name=Juniper Buffer Utilization
+report.juniper.jnxOpTableBuff.columns=jnxOpBuff
+report.juniper.jnxOpTableBuff.type=jnxOperatingTable
+report.juniper.**jnxOpTableLoad**.propertiesValues=jnxOpDescr

Could you fix it quickly? 

Sorry for the inconvenience

Cyr

**Comment by indigo423** [comment]
@StCyr There are two possibilities, you push it so your HZN-372 or just go in GitHub to the file

https://github.com/OpenNMS/opennms/blob/develop/opennms-base-assembly/src/main/filtered/etc/snmp-graph.properties.d/juniper-graph.properties

And edit the file, it will create automatically a new pull request.

**Pull request merged**